KAFKA-14499: [7/7] Add integration tests for OffsetCommit API and OffsetFetch API

dajac opened the pull request:
This patch adds integration tests for the OffsetCommit API and the OffsetFetch API. The tests runs against the old and the new group coordinator and with the new and the old consumer rebalance protocol.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
Add integration tests for the OffsetCommit API

dajac committed:
Add integration tests for the OffsetFetch API

dajac committed:
cleanup

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -55,28 +55,48 @@ public Builder(String groupId,
+            );
+        }
+
+        public Builder(String groupId,
```
Wire the member id/epoch in the builder.

dajac commented on clients/src/main/resources/common/message/OffsetFetchRequest.json:
```diff
@@ -56,7 +56,7 @@
       { "name": "GroupId", "type": "string", "versions": "8+", "entityType": "groupId",
         "about": "The group ID."},
-      { "name": "MemberId", "type": "string", "versions": "9+", "nullableVersions": "9+", "default": null, "ignorable": true,
+      { "name": "MemberId", "type": "string", "versions": "9+", "nullableVersions": "9+", "default": "null", "ignorable": true,
```
Found this bug with the tests. Yeah!

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1494,6 +1494,8 @@ class KafkaApis(val requestChannel: RequestChannel,
       requestContext,
       new OffsetFetchRequestData.OffsetFetchRequestGroup()
         .setGroupId(offsetFetchRequest.groupId)
+        .setMemberId(offsetFetchRequest.memberId)
+        .setMemberEpoch(offsetFetchRequest.memberEpoch)
```
Also found this one. Yeah!

dajac committed:
fix

dajac commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -14,233 +14,532 @@
  * limitations under the License.
  */
-
 package kafka.server
```
I rewrote all this test suite. The new one covers all the old cases as well, I think.

dajac committed:
fixup

jolshan commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -55,28 +55,48 @@ public Builder(String groupId,
                     String topicName = topicPartition.topic();
-                    OffsetFetchRequestTopic topic = offsetFetchRequestTopicMap.getOrDefault(
-                        topicName, new OffsetFetchRequestTopic().setName(topicName));
+                    OffsetFetchRequestTopics topic = offsetFetchRequestTopicMap.getOrDefault(
```
If this object represents one topic, does it make sense for the name to be plural?

dajac replied:
Yeah, this does not make sense but this is how the schema is defined. I can try to come up with a better name but I would do it separately.

jolshan commented on core/src/test/scala/unit/kafka/server/GroupCoordinatorBaseRequestTest.scala:
```diff
@@ -0,0 +1,158 @@
+import scala.jdk.CollectionConverters._
+import scala.reflect.ClassTag
+
+class GroupCoordinatorBaseRequestTest(cluster: ClusterInstance) {
```
Is the goal here to create a new baseRequestTest class? I noticed we didn't extend the existing BaseRequestTest class

dajac replied:
We can't reuse `BaseRequestTest` when using `@ClusterTest`. Here, I created `GroupCoordinatorBaseRequestTest` to group the common functionalities required to test Group Coordinator Apis.

jolshan replied:
I see. The naming reminds me of BaseRequest test and we have the same/similar `connectAndReceive` method. I think this is fine, but these test files do start to get confusing when we have so many similarly named but very different classes. Not something we need to address in this PR though.

dajac replied:
Yeah, I agree. I was considering using a test context class instead of using a base class. I am not sure if it would really be better though... What do you think?

jolshan replied:
I think it is ok as it is. I wonder if naming helps? Either way we can do as a followup

jolshan commented on core/src/test/scala/unit/kafka/server/OffsetCommitRequestTest.scala:
```diff
@@ -0,0 +1,163 @@
+  @ClusterTest(serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.consumer.max.session.timeout.ms", value = "600000"),
```
Is this the current way we enable setting new consumer group protocol? Setting these two configs?

dajac replied:
Yeah, at the moment, you have to use `unstable.api.versions.enable` and `group.coordinator.new.enable`. This will change when we release it though. It will be based on the metadata version.

jolshan replied:
Sorry I didn't ask this clearly.  For `testOffsetCommitWithNewConsumerGroupProtocolAndNewGroupCoordinator` and `testOffsetCommitWithOldConsumerGroupProtocolAndNewGroupCoordinator` the only difference was the timeout configs:

```
new ClusterConfigProperty(key = "group.consumer.max.session.timeout.ms", value = "600000")
new ClusterConfigProperty(key = "group.consumer.session.timeout.ms", value = "600000")
```

Is this how we enable the new group protocol? Both methods set `unstable.api.versions.enable` and `group.coordinator.new.enable` to be true. Should the second one have the api versions be false?

dajac replied:
The two settings that you mentioned are regular configs for the new protocol. They don't enable the new protocol. As I said in my previous comment, the new protocol is automatically enabled when `unstable.api.versions.enable` and `group.coordinator.new.enable` are set at the moment. Later, the metadata version will be used.

jolshan replied:
I feel like maybe I'm missing something obvious but the test is called `testOffsetCommitWithOldConsumerGroupProtocolAndNewGroupCoordinator()` so it says we use the old protocol.

But 
> the new protocol is automatically enabled when unstable.api.versions.enable and group.coordinator.new.enable are set at the moment.

So should one of them be false so we have the old consumer group protocol?

dajac replied:
```
  def testOffsetCommitWithOldConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
    testOffsetCommit(false)
  }
```

Whether the new or the old protocol is used is defined by the parameter passed to `testOffsetCommit`. It is not used automatically because it is available. I need to set `unstable.api.versions.enable` here to test the unreleased versions of the offset commit/fetch apis. I know.. This is confusing...

jolshan replied:
Ok -- got it. Thanks for clarifying. That's what I was looking for.

jolshan commented on core/src/test/scala/unit/kafka/server/OffsetCommitRequestTest.scala:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    // Start from version 1 because version 0 goes to ZK.
+    for (version <- 1 to ApiKeys.OFFSET_COMMIT.latestVersion(isUnstableApiEnabled)) {
```
Should do we want to consider the empty group ID case here?

dajac replied:
I can add it. It is similar for the `unknown` case.

jolshan replied:
Ah sorry -- maybe I was a bit unclear here too. Can we have the "null" group ID in this case? Or do we never have this case?

dajac replied:
Got it. `GroupId` is not nullable so that would throw a NPE on the client side. This is the reason why I did not test it. This does not bring any value from an integration test point of view.

rreddy-22 commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -14,233 +14,532 @@
+      numPartitions = 3
+    )
+
+    // Create groups and commits offsets.
```
nit: 'commit' or 'creates'

jolshan commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -14,233 +14,532 @@
+        )
+      )
+
+      assertEquals(
```
nit: comment here this is the unknown group test case.

jolshan commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -14,233 +14,532 @@
+    testMultipleGroupsOffsetFetch(useNewProtocol = false, requireStable = true)
+  }
+
+  private def testSingleGroupOffsetFetch(useNewProtocol: Boolean, requireStable: Boolean): Unit = {
```
This method is a little large and contains quite a few smaller test cases within. Is there a way we can break it up a bit more?

dajac replied:
Sure. I can separate the fetch offsets vs fetch all offsets cases.

rreddy-22 commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -14,233 +14,532 @@
+    for (version <- 8 to ApiKeys.OFFSET_FETCH.latestVersion(isUnstableApiEnabled)) {
+      assertEquals(
+        List(
+          // Fetch foo-0, foo-1 and foo-5.
```
Since we're already checking offset fetch for foo-0 and foo-1 in the next fetchAllPartitions block, did we wanna check for a mix of existing and non existent partitions?, can we also comment fetch non-existent partition foo-5

jolshan commented on core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala:
```diff
@@ -14,233 +14,532 @@
-    groups.foreach { groupId =>
-      consumerConfig.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId)
-      commitOffsets(partitionMap(groupId))
+  private def testMultipleGroupsOffsetFetch(useNewProtocol: Boolean, requireStable: Boolean): Unit = {
```
Another "my understanding" question. 😅 Do only admin clients request multiple group offsets at once? I probably should have realized this earlier, but I see we don't pass member id or epoch in this version of the request.

dajac replied:
Yeah, that's correct.

dajac committed:
fix tests & minor improvements

dajac committed:
address minor comments + add more test cases

dajac committed:
address minor comments

dajac committed:
refactor

dajac commented:
@jolshan @rreddy-22 Thanks for your comments. I have addressed all of them if not mistaken. I have also fixed a few tests and extracted common code into the base class. Please have a second look a it.

dajac committed:
refactor

jolshan reviewed (approved):
Thanks!
